### PR TITLE
RM-7530: Catch and log InvalidOperationException while pulling Docker image

### PR DESCRIPTION
**Pull request by raffaelfoidl**
https://re-motion.atlassian.net/browse/RM-7530

### CODE REVIEW

**Review by iam-mholle — APPROVED** [review]

**Review by MichaelKetting — CHANGES_REQUESTED** [review]
This design uses the IDockerClient like a 3rd party API with a pre-defiend contract. We should consider a other design-options first before catching any and all InvalidOperationExceptions thrown by this code. Please make suggestions :)

**Comment by iam-mholle** [comment]
> 
> 
> This design uses the IDockerClient like a 3rd party API with a pre-defiend contract. We should consider a other design-options first before catching any and all InvalidOperationExceptions thrown by this code. Please make suggestions :)

Valid point with the `InvalidOperationException`, the client should throw a more precise exception that can be caught. Apart from that, I don't really see the issue, retrying a pull/ignoring the error is not part of the responsibility of the simple `docker pull` abstraction the method represents, that's something the user of it should do. What about a decorator ignoring the error the pull throws?

**Comment by MichaelKetting** [comment]
> 
> 
> > This design uses the IDockerClient like a 3rd party API with a pre-defiend contract. We should consider a other design-options first before catching any and all InvalidOperationExceptions thrown by this code. Please make suggestions :)
> 
> Valid point with the `InvalidOperationException`, the client should throw a more precise exception that can be caught. Apart from that, I don't really see the issue, retrying a pull/ignoring the error is not part of the responsibility of the simple `docker pull` abstraction the method represents, that's something the user of it should do. What about a decorator ignoring the error the pull throws?

re decorator: perhaps. 
custom exception: yes, that's one way to go.

I was also considering a result (success/failed) or a parameter (fail quietly). Sometimes, I'm still unmanaged C :D But yes, a custom excpetion is probably okay for this.

File: Remotion/Web/Development.WebTesting/HostingStrategies/DockerHosting/DockerClientException.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+//
+using System;
+
+namespace Remotion.Web.Development.WebTesting.HostingStrategies.DockerHosting
+{
+  /// <summary>
+  /// The exception that is thrown when a docker command fails.
+  /// </summary>
+  public class DockerClientException : Exception
```
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/HostingStrategies/DockerHosting/DockerClientException.cs:24]: DockerOperationException, inlcude the result-code in the exception as separate field.

File: Remotion/Web/Development.WebTesting/HostingStrategies/IisDockerContainerWrapper.cs [hunk 2]
```diff
@@ -49,7 +52,14 @@ public class IisDockerContainerWrapper : IDisposable
...
     public void Run ()
     {
-      _docker.Pull (_configurationParameters.DockerImageName);
+      try
+      {
+        _docker.Pull (_configurationParameters.DockerImageName);
+      }
+      catch (DockerClientException ex)
+      {
+        s_log.Error ("Pulling docker image failed", ex);
```
Comment (MichaelKetting) [inline Remotion/Web/Development.WebTesting/HostingStrategies/IisDockerContainerWrapper.cs:61]: log the image name. 
end sentence with a .
log fallback behavior.

**Review by MichaelKetting — CHANGES_REQUESTED** [review]

**Review by MichaelKetting — APPROVED** [review]

**Comment by iam-mholle** [comment]
Added to master with 03eba1708ba00b7b24c8e1ae31129a21b8d292df.

**Pull request closed without merging**